### PR TITLE
Dyn 4159 undo selection

### PR DESCRIPTION
**Pull request by filipeotero**
### Purpose

This PR contains a bug fix in which the nodes were being selected at the same time as it was being dragged. 
To differentiate the two actions, now the nodes are being recorded only if the position of the node is changed.

DeselectionUndoBugFix

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@QilongTang 

### FYIs

@jesusalvino

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs [hunk 1]
```diff
@@ -129,13 +129,14 @@ internal void BeginDragSelection(Point2D mouseCursor)
...
-            RecordSelectionForUndo();
+            //RecordSelectionForUndo();
             draggedNodes.Clear();
             foreach (ISelectable selectable in DynamoSelection.Instance.Selection)
             {
                 ILocatable locatable = selectable as ILocatable;
+                ModelBase nodeModel = selectable as ModelBase;
                 if (null != locatable)
-                    draggedNodes.Add(new DraggedNode(locatable, mouseCursor));
+                    draggedNodes.Add(new DraggedNode(locatable, mouseCursor, nodeModel.GUID));
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:139]: Maybe check null for both before referencing?

File: src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs [hunk 2]
```diff
@@ -387,11 +414,14 @@ public class DraggedNode
...
             /// 
-            public DraggedNode(ILocatable locatable, Point2D mouseCursor)
+            public DraggedNode(ILocatable locatable, Point2D mouseCursor, Guid guid)
             {
                 this.locatable = locatable;
                 deltaX = mouseCursor.X - locatable.X;
                 deltaY = mouseCursor.Y - locatable.Y;
+                initialPostiionX = locatable.X;
+                initialPostiionY = locatable.Y;
+                this.guid = guid;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:424]: Would you explain what the guid is set for?

File: src/DynamoCore/Models/DynamoModelCommands.cs [hunk 3]
```diff
@@ -265,6 +265,7 @@ private void SelectModelImpl(SelectModelCommand command)
                     if (!command.Modifiers.HasFlag(ModifierKeys.Shift) && command.ModelGuids.Count() == 1)
                         ClearSelectionAndRecordUndo();
 
+                    WorkspaceModel.RecordModelsForModification(new List<ModelBase>() { model }, CurrentWorkspace.UndoRecorder);
```
Comment (QilongTang) [inline src/DynamoCore/Models/DynamoModelCommands.cs:268]: It seems in the future everytime we add model to `Selection`, we will need to record it for Undo? If Yes, maybe it is worth to wrap this and the following line into a utility function, similar to the `ClearSelectionAndRecordUndo`.

File: src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs (hunk 2 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:424]: If `locatable` can be cast to ModelBase here, maybe guid can be retrived already without adding as a param?

File: src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs [hunk 4]
```diff
@@ -387,11 +414,14 @@ public class DraggedNode
             /// be moved. However, the movement of ILocatable will be limited by 
             /// region and that it cannot be moved beyond the region.</param>
             /// 
-            public DraggedNode(ILocatable locatable, Point2D mouseCursor)
+            public DraggedNode(ILocatable locatable, Point2D mouseCursor, Guid guid)
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:417]: This is API breaking BTW

File: src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs (hunk 1 above)
Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:139]: I will remove this as a parameter and cast it in the constructor. Thanks!

File: src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs (hunk 2 above)
Comment (filipeotero) [reply src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:424]: This `guid `is to filter the models by its guid when we get the selected nodes/annotations. Like this:
```
ModelBase model = DynamoSelection.Instance.Selection.
                        Where((x) => (x is ModelBase)).Cast<ModelBase>().FirstOrDefault(x => x.GUID == draggedNode.guid);

                    changedPositionModels.Add(model);
```
Now I'm casting it in the constructor, thanks!

File: src/DynamoCore/Models/DynamoModelCommands.cs (hunk 3 above)
Comment (filipeotero) [reply src/DynamoCore/Models/DynamoModelCommands.cs:268]: Yes, we need this. Ok, I will set this up. Thanks!

File: src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs [hunk 5]
```diff
@@ -129,11 +129,11 @@ internal void BeginDragSelection(Point2D mouseCursor)
             // Here each node in the selection is being recorded for undo right
             // before they get updated by the drag operation.
             // 
-            RecordSelectionForUndo();
-            draggedNodes.Clear();
+            //RecordSelectionForUndo();
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:132]: Let's delete this if not needed

File: src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs [hunk 6]
```diff
@@ -338,11 +340,33 @@ private void RecordSelectionForUndo()
...
-            List<ModelBase> models = DynamoSelection.Instance.Selection.
-                Where((x) => (x is ModelBase)).Cast<ModelBase>().ToList<ModelBase>();
 
-            WorkspaceModel.RecordModelsForModification(models, Model.UndoRecorder);
+            //List<ModelBase> models = DynamoSelection.Instance.Selection.
+            //    Where((x) => (x is ModelBase)).Cast<ModelBase>().ToList<ModelBase>();
+
+            //WorkspaceModel.RecordModelsForModification(models, Model.UndoRecorder);
+
+            //DynamoViewModel.RaiseCanExecuteUndoRedo();
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:349]: Same here, although we can wait until PR checks pass

File: src/DynamoCore/Models/DynamoModelCommands.cs [hunk 7]
```diff
@@ -265,7 +265,7 @@ private void SelectModelImpl(SelectModelCommand command)
                     if (!command.Modifiers.HasFlag(ModifierKeys.Shift) && command.ModelGuids.Count() == 1)
                         ClearSelectionAndRecordUndo();
 
-                    DynamoSelection.Instance.Selection.AddUnique(model);
+                    AddSelectionAndRecordUndo(model);
```
Comment (zeusongit) [inline src/DynamoCore/Models/DynamoModelCommands.cs:268]: we can remove extra spaces here

**Review by QilongTang — COMMENTED** [review]
Looks good, a few additional comments

File: src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs [hunk 8]
```diff
@@ -129,11 +129,11 @@ internal void BeginDragSelection(Point2D mouseCursor)
...
             // before they get updated by the drag operation.
             // 
-            RecordSelectionForUndo();
-            draggedNodes.Clear();
+            //RecordSelectionForUndo();
+             draggedNodes.Clear();
             foreach (ISelectable selectable in DynamoSelection.Instance.Selection)
             {
-                ILocatable locatable = selectable as ILocatable;
+                ILocatable locatable = selectable as ILocatable;
```
Comment (zeusongit) [inline src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:136]: same, spaces

**Review by reddyashish — APPROVED** [review]

**Review by QilongTang — APPROVED** [review]

**Comment by QilongTang** [comment]
@zeusongit any more comments on this one?

**Review by zeusongit — APPROVED** [review]

**Pull request merged**